text/0080-datetime-extension.md

```diff
@@ -0,0 +1,484 @@
+
+The `datetime` type does not provide a way for a policy author to create a `datetime` from a numeric timestamp. One of the readable formats listed above must be used, instead.
+
+Values of type `datetime` have the following methods:
+
+- `.offset(duration)` returns a new `datetime`, offset by duration.
+- `.toDate()` returns a new `datetime`, truncating to the day, such that printing the `datetime` would have `00:00:00` as the time.
+- `.toTime()` returns a new `duration`, removing the days, such that only milliseconds since `.toDate()` are left. This is equivalent to `DT - DT.toDate()`
+
+Values of type `datetime` can also be used with the `-` operator.
+
+- `DT1 - DT2` returns the difference between `DT1` and `DT2` as a `duration`. (The inverse of `-` is `DT1.offset(duration)`)
+
+An invariant for the `-` operator is that when `DT1` is before `DT2` the resulting duration is negative.
+
+`DT1` may be negative (_i.e._ milliseconds _before_ `1970-01-01T00:00:00Z`) requiring that `-` be computed specially:
+
+```
+if DT1 < 0 && DT2 >= 0 then
+    result = -(DT1 - DT2)
+else
+    result = DT1 - DT2
+```
+
+There is not a use case for adding, or multiplying `datetime` values together. As such, `DT1 * DT2` and `DT1 + DT2` are errors.
+
+Values of type `datetime` can be used with comparison operators:
+
+- `DT1 < DT2` returns `true` when `DT1` is before `DT2`
+- `DT1 <= DT2` returns `true` when `DT1` is before or equal to `DT2`
+- `DT1 > DT2` returns `true` when `DT1` is after `DT2`
+- `DT1 >= DT2` returns `true` when `DT1` is after or equal to `DT2`
+- `DT1 == DT2` returns `true` when `DT1` is equal to `DT2`
+- `DT1 != DT2` returns `true` when `DT1` is not equal to `DT2`
+
+Equality is based on the underlying representation (see below) so, for example, `datetime("2024-08-21T") == datetime("2024-08-21T00:00:00.000Z")` is true. This behavior is consistent with the decimal extension function, where `decimal("1.0") == decimal("1.0000")` is also true.
+
+
+#### Representation
+
+The `datetime` type is internally represented as a `long` and contains a Unix Time in milliseconds. This is the number of non-leap seconds that have passed since `1970-01-01T00:00:00Z` in milliseconds. Unix Time days are always 86,400 seconds and handle leap seconds by absorbing them at the start of the day. Due to using Unix Time, and not providing a "current time" function, Cedar avoids the complexities of leap second handling, pushing them to the system and application.
+
+Negative Unix Time values represent the number of milliseconds before `1970-01-01T00:00:00Z`. 
+
+### Durations of Time (`duration`)
+
+The `duration(long, string)` function constructs a duration value. The string argument must be one of `"days", "hours", "minutes", "seconds", "milliseconds"`. Strict validation requires `long` and `string` to be literals, although evaluation/authorization support any appropriately-typed expressions. Values of type `duration` have the following methods:
```

Comment:
Technically, strict validation doesn’t need to require the `long` and `string` in this case to be literals.  And that’s because it’s possible to encode the meaning of this constructor to SMT for arbitrary (i.e., unknown) values.

We require arguments to `ip` and `decimal` to be literals only because it would be too expensive / infeasible to encode their behavior (parsing) on arbitrary (unknown) strings.

That said, we may want to make these literals anyway for consistency with other operators.  And, admittedly, making them literals for the purpose of strict validation would help make the analysis more efficient.